table/tables/tables.go

```diff
@@ -323,6 +336,74 @@ func (t *Table) rebuildIndices(ctx context.Context, h int64, touched []bool, old
 	return nil
 }
 
+// check unique indices constrains
+func (t *Table) checkCreateUnqIdxAvail(txn kv.Transaction, r []interface{}, h int64) (duplicate bool, recordID int64, err error) {
+	for _, v := range t.indices {
+		if v == nil {
+			continue
+		}
+		colVals, _ := v.FetchValues(r)
+		duplicate, err = v.X.CheckDuplicated(txn, colVals, h)
+		if err != nil {
+			return false, 0, errors.Trace(err)
+		}
+
+		if duplicate {
+			recordID, err = getDuplicateRow(txn, colVals, v)
+			if err != nil {
+				return duplicate, 0, errors.Trace(err)
+			}
+			return duplicate, recordID, nil
+		}
+	}
+
+	return false, 0, nil
+}
+
+func isIndexValsEqual(vals1, vals2 []interface{}) bool {
+	byteVals1, _ := kv.EncodeValue(vals1...)
+	byteVals2, _ := kv.EncodeValue(vals2...)
+
+	return bytes.Equal(byteVals1, byteVals2)
+}
+
+func (t *Table) checkUpdateUnqIdxAvail(txn kv.Transaction, oldData, newData []interface{}, h int64) (duplicate bool, err error) {
+	for _, v := range t.indices {
+		if v == nil {
+			continue
+		}
+		oldVals, _ := v.FetchValues(oldData)
+		newVals, _ := v.FetchValues(newData)
+		if isIndexValsEqual(oldVals, newVals) { // update the same index value is not necessary but available
+			continue
+		}
+
+		duplicate, err = v.X.CheckDuplicated(txn, newVals, h)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+
+		if duplicate {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func getDuplicateRow(txn kv.Transaction, colVals []interface{}, v *column.IndexedCol) (recordID int64, err error) {
```

Comment:
CheckDuplicated may return handle directly at same time, so we don't need to call getDuplicatRow again.